spec/sanctorale_writer_spec.rb

```diff
@@ -0,0 +1,156 @@
+require 'spec_helper'
+
+describe CR::SanctoraleWriter do
+  let(:s) { CR::Sanctorale.new }
+  let(:d) { CR::SanctoraleWriter.new }
+
+  describe 'record/file format' do
+    let(:celebration) do
+      CR::Celebration.new(
+        title: 'The Most Holy Name of Jesus',
+        symbol: :name_jesus,
+        date: CR::AbstractDate.new(1, 3)
+      )  
+    end
+    let(:result) { d.send(:celebration_line, celebration) }
+
+    describe 'day' do
+      it 'prints the day' do
+        expect(result[0]).to eq('3')
+      end
+    end
+
+    describe 'title' do
+      it 'prints the title' do
+        expect(result).to include(': The Most Holy Name of Jesus')
+      end
+    end
+
+    describe 'symbol' do
+      it 'prints the symbol' do
+        expect(result).to include(' name_jesus :')
+      end
+    end
+
+    describe 'rank' do
+      describe 'memorial' do
+        let(:celebration) do
+          CR::Celebration.new(
+            rank: CR::Ranks::MEMORIAL_GENERAL,
+            date: CR::AbstractDate.new(7, 26)
+          )  
+        end
+        it 'includes m as rank' do
+          expect(result).to include(' m ')
+        end
+      end
+      describe 'feast' do
+        let(:celebration) do
+          CR::Celebration.new(
+            rank: CR::Ranks::FEAST_GENERAL,
+            date: CR::AbstractDate.new(9, 29)
+          )  
+        end
+        it 'includes f as rank' do
+          expect(result).to include(' f ')
+        end
+      end
+      describe 'solemnity' do
+        let(:celebration) do
+          CR::Celebration.new(
+            rank: CR::Ranks::SOLEMNITY_GENERAL,
+            date: CR::AbstractDate.new(12, 8)
+          )  
+        end
+        it 'includes s as rank' do
+          expect(result).to include(' s ')
+        end
+      end
+
+      describe 'feast of the lord' do
+        let(:celebration) do
+          CR::Celebration.new(
+            rank: CR::Ranks::FEAST_LORD_GENERAL,
+            date: CR::AbstractDate.new(8, 6)
+          )  
+        end
+        it 'includes f2.5 as rank' do
+          expect(result).to include(' f2.5 ')
+        end
+      end
+    end
+    
+    describe 'colour' do
+      describe 'red' do
+        let(:celebration) do
+          CR::Celebration.new(
+            colour: CR::Colours::RED,
+            date: CR::AbstractDate.new(8, 14)
+          )  
+        end
+        it 'includes R as colour' do
+          expect(result).to include(' R ')
+        end
+      end
+    end
+  end
+
+  describe 'exceptional cases' do
+    it 'raises an exception if no date is provided' do
+      c = CR::Celebration.new
+      expect { d.send(:celebration_line, c) }.to raise_exception(NoMethodError)
```

Comment:
While trying to remove unnecessary fields, I found a small issue here in the case where no date is present. I think we can handle it better, but I'm not sure what the best approach is. Should a date be a required parameter (i.e. should the `Celebration` constructor raise an exception if `date` is not present)? Should we raise an error during serialization if we try to serialize a `Celebration` without a date?

Reply:
Ouch, in my code review I missed that in `#celebration_line` you use `Celebration#date` as the date of the celebration to be serialized. That's not correct. While it may work in "usual" cases, values of `Celebration#date` are not part of `Sanctorale`'s contract. `Sanctorale` does not promise that all `Celebration`s it returns have non-nil `#date`, nor that the `#date` (if non-nil) is the same as the date the `Celebration` happens to occur on in this particular `Sanctorale` instance.

Please change `#celebration_line` to serialize the date used as key in `Sanctorale` (and yielded by `#each_day`) instead.

Reply:
Done. Lucky for me, this was a pretty simple adjustment :slightly_smiling_face: